File: engine/algo.js
// algo.js

class Algo {
    constructor(deps) {
        this.deps = deps;
        
        this.neutCost = 2;
        this.markedPixels = [];
    }
  
    attackProcessInit() {
        this.markPossiblePixels();
        if (this.markedPixels.length === 0) {
            this.returnRemaining();
        } else {
            const remaining = this.deps.speed.remaining;
            if (Math.floor(remaining / this.markedPixels.length) > this.neutCost) {
                this.takeBorderPixels();
            } else {
                this.returnRemaining();
            }
        }
    }
  
    markPossiblePixels() {
        this.markedPixels = [];
        const borderPixels = this.deps.pixel.getBorderPixels();
        for (let borderPixel of borderPixels) {
            for (let side = 0; side <= 3; side++) {
                let x = this.deps.pixel.getX(borderPixel),
                    y = this.deps.pixel.getY(borderPixel);
                switch (side) {
                    case 0:
                        x--;
                        break;
                    case 1:
                        x++;
                        break;
                    case 2:
                        y--;
                        break;
                    case 3:
                        y++;
                        break;
                }
                const pIndex = this.deps.pixel.getIndex(x, y);
                if (this.deps.pixel.mapArray[pIndex] === 0 && !this.markedPixels.includes(pIndex)) {
                    this.markedPixels.push(pIndex);
                }
            }
        }
    }
  
    returnRemaining() {
        this.deps.interest.troops += this.deps.speed.remaining;
        this.deps.gameStatistics.expenses[1] -= this.deps.speed.remaining;
        this.deps.speed.removeEntry();
    }
  
    takeBorderPixels() {
        this.deps.speed.remaining -= this.markedPixels.length * this.neutCost;
        for (let pIndex of this.markedPixels) {
            this.deps.pixel.mapArray[this.deps.pixel.getIndex(this.deps.pixel.getX(pIndex), this.deps.pixel.getY(pIndex))] = 2;
        }
        this.deps.pixel.updatePixels();
    }
}

module.exports = Algo;

File: engine/algo.rs
use crate::pixel::Pixel;
use crate::speed::Speed;
use crate::interest::Interest;
use crate::game_statistics::GameStatistics;
use serde::Serialize;

/// Serializable state for the Algo component
#[derive(Debug, Clone, Serialize)]
pub struct AlgoState {
    pub neut_cost: i32,
    pub marked_pixel_count: i32,
}

/// Attack algorithm for territorial expansion
/// 
/// This is a simplified version of the original implementation.
/// The original had complex pixel marking and border detection logic
/// using a 2D map array with boundary tracking.
#[derive(Debug, Clone)]
pub struct Algo {
    /// Cost in troops to neutralize one pixel (neut cost)
    neut_cost: i32,
    /// Number of pixels that can be attacked this round
    marked_pixel_count: i32,
}

impl Algo {
    pub fn new() -> Self {
        Self {
            neut_cost: 2,
            marked_pixel_count: 0,
        }
    }

    /// Initialize attack processing
    /// Decides whether to attack border pixels or return remaining troops
    pub fn attack_process_init(&mut self, speed: &mut Speed, pixel: &mut Pixel, interest: &mut Interest, game_statistics: &mut GameStatistics) {
        self.marked_pixel_count = pixel.border + pixel.border_increment;
        if self.marked_pixel_count == 0 {
            self.return_remaining(speed, interest, game_statistics);
        } else {
            let remaining = speed.remaining;
            if remaining / self.marked_pixel_count > self.neut_cost {
                self.take_border_pixels(speed, pixel);
            } else {
                self.return_remaining(speed, interest, game_statistics);
            }
        }
    }

    /// Return unused troops to the interest pool
    /// Also refunds the attack expense since troops weren't used
    pub fn return_remaining(&self, speed: &mut Speed, interest: &mut Interest, game_statistics: &mut GameStatistics) {
        interest.troops += speed.remaining;
        game_statistics.expenses[1] -= speed.remaining;
        speed.remove_entry();
    }

    /// Execute the attack on border pixels
    /// Consumes troops to gain land and expand the border
    pub fn take_border_pixels(&self, speed: &mut Speed, pixel: &mut Pixel) {
        speed.remaining -= self.marked_pixel_count * self.neut_cost;
        pixel.land += self.marked_pixel_count;
        pixel.border += pixel.border_increment;
    }

    /// Load state from a previous simulation
    pub fn load_state(&mut self, state: &AlgoState) {
        self.neut_cost = state.neut_cost;
        self.marked_pixel_count = state.marked_pixel_count;
    }

    /// Save current state for serialization
    pub fn save_state(&self) -> AlgoState {
        AlgoState {
            neut_cost: self.neut_cost,
            marked_pixel_count: self.marked_pixel_count,
        }
    }
}